File: modules/action/src/main/scala/action/common/ActionConsumer.scala
package action.common

import java.util.UUID

import model.messages.{ActionRequest, ActionResponse}
import model.specs.ActionProcessorSpec
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.kstream.{Consumed, KStream}
import topics.topics.{TopicNamer, TopicTypes}

object ActionConsumer {

  def actionRequestStream[A](spec: ActionProcessorSpec[A],
                             actionTopicNamer: TopicNamer,
                             builder: StreamsBuilder): KStream[UUID, ActionRequest[A]] =
    builder.stream[UUID, ActionRequest[A]](
      actionTopicNamer(TopicTypes.ActionTopic.request),
      Consumed.`with`(spec.serdes.uuid, spec.serdes.request)
    )

  def actionResponseStream[A](spec: ActionProcessorSpec[A],
                              actionTopicNamer: TopicNamer,
                              builder: StreamsBuilder): KStream[UUID, ActionResponse] =
    builder.stream[UUID, ActionResponse](
      actionTopicNamer(TopicTypes.ActionTopic.response),
      Consumed.`with`(spec.serdes.uuid, spec.serdes.response)
    )
}

File: modules/action/src/main/java/io/simplesource/saga/action/common/ActionConsumer.java
package io.simplesource.saga.action.common;

import java.util.UUID;

import io.simplesource.saga.action.sourcing.SourcingStream;
import io.simplesource.saga.model.messages.ActionRequest;
import io.simplesource.saga.model.messages.ActionResponse;
import io.simplesource.saga.model.specs.ActionProcessorSpec;
import io.simplesource.saga.shared.topics.TopicNamer;
import io.simplesource.saga.shared.topics.TopicTypes;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.kstream.Consumed;
import org.apache.kafka.streams.kstream.KStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


public class ActionConsumer {
    static Logger logger = LoggerFactory.getLogger(ActionConsumer.class);

    public static <A> KStream<UUID, ActionRequest<A>> actionRequestStream(
            ActionProcessorSpec<A> spec,
            TopicNamer actionTopicNamer,
            StreamsBuilder builder) {
        return builder.stream(
                actionTopicNamer.apply(TopicTypes.ActionTopic.request),
                Consumed.with(spec.serdes.uuid(), spec.serdes.request())
        ).peek(Utils.logValues(logger, "actionRequestStream"));
    }

    public static <A> KStream<UUID, ActionResponse> actionResponseStream(ActionProcessorSpec<A> spec,
                                                                  TopicNamer actionTopicNamer,
                                                                  StreamsBuilder builder) {
        return builder.stream(
                actionTopicNamer.apply(TopicTypes.ActionTopic.response),
                Consumed.with(spec.serdes.uuid(), spec.serdes.response())
        ).peek(Utils.logValues(logger, "actionResponseStream"));
    }
}
